pkg/speaker/vip/keepalived.go

```diff
@@ -1,334 +1,344 @@
+	})
 
-	if reflect.DeepEqual(cm.Data, k.data) {
-		return nil
+	klog.Infof("nodes %s", strings.Join(nodenames, ","))
+	bytes := sha256.Sum256([]byte(strings.Join(nodenames, ",")))
+	return bytes
+}
+
+// isNodeInList returns true if the node is in the nodes list
+func (k *KeepAlived) isNodeInList(nodes []corev1.Node) bool {
+	for _, node := range nodes {
+		if node.Name == util.GetNodeName() {
+			return true
+		}
 	}
 
-	cm.Data = k.data
-	_, err = k.client.CoreV1().ConfigMaps(util.EnvNamespace()).Update(ctx, cm, metav1.UpdateOptions{})
-	return err
+	return false
 }
 
-func (k *KeepAlived) Start(stopCh <-chan struct{}) error {
-	cm, err := k.getAndCreateConfig()
-	if err != nil {
-		k.log.Error(err, fmt.Sprintf("get or create cm:%s error", cm.Name))
-		return err
+func (k *KeepAlived) DelBalancer(vip string) error {
+	instanceName, exist := k.vips[vip]
+	if !exist {
+		return nil
 	}
+	delete(k.vips, vip)
 
-	dsClient := k.client.AppsV1().DaemonSets(util.EnvNamespace())
-	vipds, err := dsClient.Get(context.TODO(), constant.OpenELBVipName, metav1.GetOptions{})
-	if err != nil {
-		if !errors.IsNotFound(err) {
-			k.log.Error(err, "keepalived daemonSet get error")
-			return err
-		}
+	instance, exist := k.instances[instanceName]
+	if !exist {
+		return nil
+	}
+	delete(k.instances, instanceName)
+	k.idAlloc.Free(instance.RouteID)
 
-		// not found daemonset
-		vipds, err = dsClient.Create(context.TODO(), k.generateVIPDaemonSet(metav1.OwnerReference{
-			Kind:               "ConfigMap",
-			APIVersion:         "v1",
-			Name:               cm.GetName(),
-			UID:                cm.GetUID(),
-			BlockOwnerDeletion: pointer.Bool(true),
-			Controller:         pointer.Bool(true),
-		}), metav1.CreateOptions{})
-		if err != nil {
-			k.log.Error(err, "keepalived daemonSet create error")
-			return err
-		}
-		k.log.Info(fmt.Sprintf("keepalived daemonSet %s created successfully", vipds.Name))
+	if err := k.WriteCfg(); err != nil {
+		klog.Error(err)
+		return err
 	}
+	return k.Reload()
+}
 
+func (k *KeepAlived) Start(stopCh <-chan struct{}) error {
+	if err := k.WriteCfg(); err != nil {
+		klog.Error(err)
+		return err
+	}
+
+	started := make(chan bool, 1)
 	go func() {
-		select {
-		case <-stopCh:
-			deletePolicy := metav1.DeletePropagationForeground
-			deleteOpt := metav1.DeleteOptions{
-				PropagationPolicy: &deletePolicy,
+		for {
+			var logWriter io.WriteCloser
+			if k.logPath != "" {
+				logWriter = &lumberjack.Logger{
+					Filename:   k.logPath,
+					MaxSize:    100,
+					MaxBackups: 3,
+					MaxAge:     28,
+					Compress:   true,
+				}
+			} else {
+				logWriter = newKeepalivedLogPiper()
 			}
-
-			if err = k.client.CoreV1().ConfigMaps(util.EnvNamespace()).Delete(context.TODO(), cm.Name, deleteOpt); err != nil {
-				k.log.Error(err, "keepalived configMap delete error")
+			defer logWriter.Close()
+			k.cmd = exec.Command(keepalivedStarter, "--dont-fork", "--log-console", "--log-detail", "--vrrp", "--release-vips")
```

Comment:
Can we provide an external parameter that allows users to customize the parameters of keepalived?

Reply:
done.